main.go

```diff
@@ -7,103 +7,151 @@ import (
+			Pass:         password,
+			HTTPPostMode: true,
+			DisableTLS:   true,
+		}
+
+		var err2 error
+		client, err2 = rpcclient.New(connCfg, nil)
+		if err2 != nil {
+			log.Fatal(err2)
+		}
+	})
+}
+
+func clearCachePeriodically() {
+	for {
+		time.Sleep(10 * time.Minute)
+
+		cacheMutex.Lock()
+		for targetTime, entry := range cacheMap {
+			if time.Since(entry.timestamp) > cacheTTL {
+				delete(cacheMap, targetTime)
+			}
 		}
-		// fmt.Println(midBlock.Time)
+		cacheMutex.Unlock()
+	}
+}
+
+func getBlockTime(height int64) (int64, error) {
+	hash, err := client.GetBlockHash(height)
+	if err != nil {
+		return 0, fmt.Errorf("error getting block hash: %v", err)
+	}
+	block, err := client.GetBlockVerbose(hash)
+	if err != nil {
+		return 0, fmt.Errorf("error getting block details: %v", err)
+	}
+	return block.Time, nil
+}
+
+func binarySearch(blockCount int64, targetTime int64) int64 {
+	// Check if the result is already cached
+	cacheMutex.RLock()
+	if cachedEntry, ok := cacheMap[targetTime]; ok {
+		cacheMutex.RUnlock()
+		return cachedEntry.height
+	}
+	cacheMutex.RUnlock()
+
+	// Get the latest block's time
+	latestBlockHash, err := client.GetBlockHash(blockCount)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		// leftBlockHash, err := client.GetBlockHash(leftBlockHeight)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	latestBlock, err := client.GetBlockVerbose(latestBlockHash)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		// leftBlock, err := client.GetBlockVerbose(leftBlockHash)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	latestBlockTime := latestBlock.Time
 
-		// rightBlockHash, err := client.GetBlockHash(rightBlockHeight)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	// Check if the target time is in the future
+	if targetTime > latestBlockTime {
+		timeDifference := targetTime - latestBlockTime
+		estimatedFutureBlocks := timeDifference / averageBlockTime
+		return blockCount + estimatedFutureBlocks
+	}
 
-		// rightBlock, err := client.GetBlockVerbose(rightBlockHash)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	// Perform binary search for past blocks
+	var leftBlockHeight, rightBlockHeight int64 = 0, blockCount
+
+	for leftBlockHeight <= rightBlockHeight {
+		midBlockHeight := (leftBlockHeight + rightBlockHeight) / 2
 
-		midBlockTime := midBlock.Time
-		// leftBlockTime := leftBlock.Time
-		// rightBlockTime := rightBlock.Time
+		midBlockTime, err := getBlockTime(midBlockHeight)
+		if err != nil {
+			log.Printf("Error getting block time: %v", err)
+			return -1
+		}
 
 		if midBlockTime == targetTime {
+			cacheMutex.Lock()
+			cacheMap[targetTime] = cacheEntry{height: midBlockHeight, timestamp: time.Now()}
+			cacheMutex.Unlock()
 			return midBlockHeight
 		} else if midBlockTime < targetTime {
 			leftBlockHeight = midBlockHeight + 1
 		} else {
 			rightBlockHeight = midBlockHeight - 1
 		}
 	}
-	return (leftBlockHeight)
+
+	result := leftBlockHeight
+	cacheMutex.Lock()
```

Comment:
Why is it necessary to use a mutex when writing to the cache?

Reply:
I think I just found the answer to my own question: "Goroutines run in the same address space, so access to shared memory must be synchronized."  From go.dev

Reply:
Its not good to write to a shared space in parallel so mutex locks will prevent that.